DigitalLearningSolutions.Data/DataServices/SupervisorDelegateDataService.cs

```diff
@@ -73,10 +79,10 @@ public IEnumerable<SupervisorDelegate> GetPendingSupervisorDelegateRecordsByEmai
                     FROM SupervisorDelegates sd
                     INNER JOIN AdminUsers au ON sd.SupervisorAdminID = au.AdminID
                     WHERE au.CentreID = @centreId
-                      AND sd.DelegateEmail = @email
+                      AND sd.DelegateEmail IN @emails

```

Comment:
I'm not sure how Dapper implements this - could you test it? I'd expect it to work fine if `emails` is empty though...
It's possible for multiple supervisors to have invited the same delegate, so this method should return an IEnumerable.

Reply:
Ok, it looks like an empty list results in `AND sd.DelegateEmail IN (SELECT @emails WHERE 1 = 0)`, i.e. a safe but unnecessary query.

If the first item in `emails` is null, you get `System.NotSupportedException : The first item in a list-expansion cannot be null` from Dapper, so that is a problem. Looking at the usages, it should not be possible to create a list where the first entry is null, because the list is created with primary email first, then centre specific, but this is non-obvious and fluky. I therefore propose:

* Change `GetPendingSupervisorDelegateRecordsByEmailsAndCentre` so that `emails` are non-nullable strings
* Change `GetPendingSupervisorDelegateIdsMatchingDelegate` so that `delegateRegistrationModel.CentreSpecificEmail` is only added to `delegateEmails` if it is not null

Reply:
Ah I see, I figured it would be fine since the primary email will never be null so it would always be called with at least one not null value, but better to write it so we only give it non-null values. 

I've changed the service method to filter out null/empty values so we only have valid emails when we call the data service